src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc

```diff
@@ -589,76 +490,86 @@ static void pf_connectivity_changed_locked(grpc_exec_ctx *exec_ctx, void *arg,
           GRPC_CLOSURE_SCHED(exec_ctx, pp->on_complete, GRPC_ERROR_NONE);
           gpr_free(pp);
         }
-        grpc_connected_subchannel_notify_on_state_change(
-            exec_ctx, p->selected, p->base.interested_parties,
-            &p->checking_connectivity, &p->connectivity_changed);
-        break;
-      case GRPC_CHANNEL_TRANSIENT_FAILURE:
-        p->checking_subchannel =
-            (p->checking_subchannel + 1) % p->num_subchannels;
-        if (p->checking_subchannel == 0) {
-          /* only trigger transient failure when we've tried all alternatives
-           */
+        // Renew notification.
+        grpc_lb_subchannel_data_start_connectivity_watch(exec_ctx, sd);
+        return;
+      }
+      case GRPC_CHANNEL_TRANSIENT_FAILURE: {
+        do {
+          sd->subchannel_list->checking_subchannel =
+              (sd->subchannel_list->checking_subchannel + 1) %
+              sd->subchannel_list->num_subchannels;
+          sd = &sd->subchannel_list
+                    ->subchannels[sd->subchannel_list->checking_subchannel];
+        } while (sd->subchannel == NULL);
+        // Case 1: Only set state to TRANSIENT_FAILURE if we've tried
+        // all subchannels.
+        if (sd->subchannel_list->checking_subchannel == 0 &&
+            sd->subchannel_list == p->subchannel_list) {
           grpc_connectivity_state_set(
               exec_ctx, &p->state_tracker, GRPC_CHANNEL_TRANSIENT_FAILURE,
               GRPC_ERROR_REF(error), "connecting_transient_failure");
         }
+        sd->curr_connectivity_state =
+            grpc_subchannel_check_connectivity(sd->subchannel, &error);
         GRPC_ERROR_UNREF(error);
-        p->checking_connectivity = grpc_subchannel_check_connectivity(
-            p->subchannels[p->checking_subchannel], &error);
-        if (p->checking_connectivity == GRPC_CHANNEL_TRANSIENT_FAILURE) {
-          grpc_subchannel_notify_on_state_change(
-              exec_ctx, p->subchannels[p->checking_subchannel],
-              p->base.interested_parties, &p->checking_connectivity,
-              &p->connectivity_changed);
-        } else {
-          goto loop;
+        if (sd->curr_connectivity_state == GRPC_CHANNEL_TRANSIENT_FAILURE) {
+          // Reuses the connectivity refs from the previous watch.
+          grpc_lb_subchannel_data_start_connectivity_watch(exec_ctx, sd);
```

Comment:
The intent of pick_first is that we iterate through the subchannels in the order that the addresses were specified -- that's why it's called pick_first instead of pick_any. :)

Also, I think that once a channel goes into state TRANSIENT_FAILURE and we stop watching it, it will not try to connect again until we start watching it again.  So it wouldn't actually help us to check the status of any other channel here, unless we decided to try to connect to all of the subchannels in parallel (which we don't want to do in pick_first, because in the common case where all backends are reachable, we don't want to connect to more than one of them).

Reply:
There must be something I am missing..

I think if we just continue trying the current subchannel instead of the next one when both of them are `TRANSIENT_FAILURE`, the order won't be broken. 

Do we choose the next one to watch because it's been in `TRANSIENT_FAILURE` for a longer period hence is more probable to be ready soon?

Reply:
The expected behavior of the pick_first LB policy is to try to connect to the subchannels in the order they were given to us and to pick the first one for which the connection attempt succeeds.  So the logic is basically the following (pseudo-code):

```
for subchannel in subchannel_list:
  try to connect to subchannel
  if connection attempt succeeded:
    p->selected = subchannel
    break
```

When we try to connect to a given subchannel, the subchannel goes from IDLE to CONNECTING.  If the connection attempt succeeds, the subchannel goes from CONNECTING to READY, at which point we choose that subchannel and don't try to connect to any others.  However, if the connection attempt fails, the subchannel goes from CONNECTING to TRANSIENT_FAILURE, which is how we know that it failed.  When this happens, we need to move on to the next subchannel in the list and start attempting to connect to it.

If we didn't move on to the next subchannel when the current one goes into TRANSIENT_FAILURE, then we would only ever be checking the first subchannel in the list -- we would never move on to trying any of the others.

(Note that because the subchannel code is currently lying to us and converting TRANSIENT_FAILURE into SHUTDOWN, this code-path will never actually be triggered right now.  However, the same explanation applies to the SHUTDOWN case, and the logic is essentially the same there.)

Reply:
I think if the failure is really transient, we can still watch it unless it is shut down or ready. However, if we don't want to wait for a subchannel in `TRANSIENT_FAILURE`, we can continue to find the next subchannel that is not in `TRANSIENT_FAILURE`. 

But I don't quite understand why we skip the current subchannel in `TRANSIENT_FAILURE`, and watch the next subchannel in `TRANSIENT_FAILURE`. It's kind of a mix.. And the only reason I can think of is that we have more hope to the subchannel in `TRANSIENT_FAILURE` for a longer time.

Reply:
I think the intent here is that any time a subchannel transitions into TRANSIENT_FAILURE, that means that the most recent connection attempt has failed.  In other words, TRANSIENT_FAILURE is expected to always mean that a connection attempt has failed, and we won't try connecting to that subchannel again until we loop back around to it (i.e., wrap-around after iterating through the whole list).  If we did loop back around, then we would expect that on the second attempt, the channel would start in state TRANSIENT_FAILURE, because that's where we left it on the last attempt, and it will never have changed out of that state in the interim, since we have not been watching it.

That having been said, as I mentioned above, this code is not ever actually getting executed today, because the subchannel code is always lying by converting TRANSIENT_FAILURE to SHUTDOWN.  And as a result of the way we're handling SHUTDOWN, we never actually wrap around to the beginning of the list -- we only give each subchannel one attempt, and if it fails we unref it and move on to the next one.

In any case, this is basically pre-existing logic, and I did not want to change it as part of this PR.  But I do agree that we will probably want to rework this when we eventually get around to fixing the lie in the subchannel code.

Reply:
Yes, I'm fine with the "try the failed one again later" logic as long as that is the intended logic. I agree that we should put it aside for now and work on it later.